### DISCUSSION TITLE
[HELP] Can't find path when launching from Pegasus frontend

### DISCUSSION THREAD

**Discussion by SoumyBhow** [rafaelvcaetano/melonDS-android#558] [Support]
I've setup MelonDS in Pegasus frontend with the following launch code:
command: am start --user 0 
  -a android.intent.action.VIEW
  -n me.magnum.melonds/.ui.emulator.EmulatorActivity
  -e PATH "{file.path}"

However, when launching a game, MelonDS opens but it says it couldn't find the ROM path (attached pic)
![Screenshot_20210810-231541001](https://user-images.githubusercontent.com/42043870/129099684-0e7d97a8-e668-4f6f-aeeb-3380d8bb6a97.jpg)

System is Android 11 on OnePlus 7 Pro.

When replacing {file.path} with an actual ROM's directory, that game is able to launch from Pegasus, so I'm thinking it's a problem with {file.path}?
I've already made an issue on Pegasus' git (https://github.com/mmatyas/pegasus-frontend/issues/878), but I thought it wouldn't hurt to see if it's an issue on MelonDS' side.

**Comment by rafaelvcaetano** [comment]
It's quite possible that this is a melonDS issue. As the message says, the ROM you are trying to load must be available in the emulator's ROM list. I know this is stupid, but with the most recent changes to Android storage access, there's no other way to ensure that the emulator can access the ROM and save file. As such, make sure that the ROM you are trying to load through Pegasus can be found in melonDS's ROM list.

If you can find the ROM in the list, then there might be an issue converting the file path provided by Pegasus to the one used by melonDS. Again, this doesn't make much sense, but it's the way Google wants to make things from now on. If this is the case, could you send me a screenshot of your ROM list in melonDS?

**Reply by SoumyBhow** [reply]
ROMs can be seen and launched from MelonDS' ROM list. All my ROMs can be seen there, and the search directory has already been added 
![Screenshot_20210811-214831546](https://user-images.githubusercontent.com/42043870/129126092-ccd657d5-a6c8-4385-8061-8903cfaa8ee9.jpg)
![Screenshot_20210811-214844402](https://user-images.githubusercontent.com/42043870/129126097-7dd91a25-ed93-46e2-9e5f-85132fa54932.jpg)

**Comment by rafaelvcaetano** [comment]
So, I've just tested this myself and I couldn't load the ROMs with the stable Pegasus release, but it worked perfectly using the latest release (released on 2021-08-09). Which version did you test on? If you were not using the latest release, could you try it please?

**Reply by SoumyBhow** [reply]
I've been using that latest version of Pegasus all this time. I probably should've specified so my apologies. Would you mind sharing your metadata.txt or however method you used to get it to launch?

**Reply by rafaelvcaetano** [reply]
Sure. Here is the `metadata.txt` I used:
```
collection: NDS
extension: nds
launch: am start --user 0
  -a android.intent.action.VIEW
  -n me.magnum.melonds/.ui.emulator.EmulatorActivity
  -e PATH "{file.path}"
```

One of the things you can try is replace `android.intent.action.VIEW` with `me.magnum.melonds.LAUNCH_ROM`, but I don't think it will have any impact.

And just to make sure, are you using the `alpha15-26-gfce59755` version of Pegasus?

**Reply by SoumyBhow** [reply]
Nope still the same error for me. I even downloaded the latest Pegasus APK just to be sure. Maybe there's some other part of my text file that's wrong? Would you mind taking a look at my file to see if there are any differences?
[metadata.pegasus.txt](https://github.com/rafaelvcaetano/melonDS-android/files/6987521/metadata.pegasus.txt)

**Reply by SoumyBhow** [reply]
Little update: I've scraped my ROM's with Skraper on my PC instead of Skyscraper and that works. I guess this is a scraper issue then, but I still don't understand why. I guess I'll have to look into it myself. Nonetheless, thank you for your time.

Edit: looks like it was the path after all. Replacing the location of the ROM folder with a dot fixed it. The Metadata file stays in the same folder as my ROMs so just doing ./media/boxart for example works.

**Comment by SoumyBhow** [comment]
Hey there, 

I just wanted to report that MelonDS no longer works with frontends, even after the frontend updated with scoped storage support. I've tried this out with [Daijishou](https://play.google.com/store/apps/details?id=com.magneticchen.daijishou), which recently got an update for Uri support. Mupen64, Aethersx2 and PPSSPP, some examples of emulators that adapted scoped storage, and they work no problem, yet MelonDS doesn't. I've spoken with the Daijishou dev and they suggested using absolute paths along with scoped storage via a Java interface. What are your thoughts on this?

**Reply by rafaelvcaetano** [reply]
I've recently noticed that Dig also stopped working after adding support for scoped storage. I'll have to evaluate how to properly use the new frontend implementations.

**Reply by SoumyBhow** [reply]
Thanks for offering to look into this. Much appreciated

**Comment by mike-seaver** [comment]
I'm not able to make it work through ARC Browser frontend on Android 12, not sure if there are any news related to this, but the problem matches with the one described here. I know you do this for free and I'm not in position to demand nothing, I don't wan't to bother you, if it were posible I'd like to know if you're busy or you're still working on it.

**Comment by EgoBizarro** [comment]
Heya, I just wanted to report that melonDS is still not working correctly with Daijisho (don't know about other frontends). I tried all kinds of parameters, path, uri, -a android.intent.action.VIEW and me.magnum.melonds.LAUNCH_ROM... It only opens melonDS but can't find the rom to boot it.
I must be a regression, though, because someone showed me that it was working for them, but not on 0.9.5; he's gone back to 0.9.3 (and I tried 0.9.4, so I guess it got broken somewhere between those two versions).
He has it working with this:
-a android.intent.action.VIEW
-n me.magnum.melonds/.ui.emulator.EmulatorActivity
-e PATH {file.path}

**Reply by rafaelvcaetano** [reply]
It should be fixed in the latest nightly release. If you want to try it, you can find it here: https://github.com/rafaelvcaetano/melonDS-android/actions

Just click on top-most release, scroll down and you will find the APK.

**Reply by EgoBizarro** [reply]
Thank you, but I've been using that one for a few weeks and tried tying it up in Daijisho. No dice... or did you get it to work with some specific parameters?

**Reply by PiixelNerd** [reply]
I can use melonds 0.9.5 on daijisho easily, I just selected it from the list of emulators and it worked.

**Reply by EgoBizarro** [reply]
Huh? Do you have an older Android without scoped storage? That be v10 or below

**Reply by PiixelNerd** [reply]
Android 12

**Reply by EgoBizarro** [reply]
Weird... what is the launch text in your 'player' for melonDS?

**Reply by PiixelNerd** [reply]
Veja...

https://user-images.githubusercontent.com/93597875/205514290-f77a3d9e-e901-43fb-88f6-1f6cd47d6c79.mp4

**Reply by EgoBizarro** [reply]
Thank you for doing that but unfortunately that's not what I need. I think you have to go to settings -> inspect players. The important stuff looks something like this:
-a android.intent.action.VIEW
-n me.magnum.melonds/.ui.emulator.EmulatorActivity
-e PATH {file.path}

**Reply by PiixelNerd** [reply]
I'm sorry, I don't understand much about Daijisho.  But what is the advantage of this process you are trying to perform?  Wouldn't the idea be just to run melonds for him?  because the way I showed it doesn't work?

**Reply by EgoBizarro** [reply]
? That's the thing. You say it works for you with melonds 0.9.5 on daijisho, but for me (and others) it doesn't work and I'm trying to find why. That's why I want to compare my launch parameters with yours, to see what you have different.

**Reply by PiixelNerd** [reply]
I did the whole process from the beginning, I hope it helps you!  because here for me it works normally... and I didn't change any parameters in Daijisho!

https://user-images.githubusercontent.com/93597875/205517371-6d68ed64-e4b5-471e-97dd-f377132c237b.mp4

**Reply by EgoBizarro** [reply]
Thank you very much, I really appreciate it! Unfortunately, it still doesn't work for me. I also deleted the platformand player and started from scratch, but still. Nothing more I can think of, so I guess I'll just have to wait until someone figures it out. Thanks again!

**Reply by rafaelvcaetano** [reply]
Thank you guys for digging deeper to find a solution. I've also spent some time looking at potential issues, but I couldn't find anything that might be causing a problem. @EgoBizarro which device are you using? It might be a problem specific to that device or manufacturer. And do other emulators work fine?

**Reply by EgoBizarro** [reply]
I have a 'Poco X3 NFC' with Android 12. All other emulators work fine with Daijisho (except for the few that just generally aren't compatible with frontends); that's 9 emulators + RetroArch with a few cores. The roms aren't even on an external SD card or something (though I als have lots of roms on that one and that's also no problem).
I'm not the only one with the issue, though, otherwise I wouldn't have come here I guess. I think it was actually the Daijisho dev himself that told me only 0.9.3 worked for him with Daijisho.

**Reply by PiixelNerd** [reply]
well, I also have no idea what it could be, I have android 12 (Galaxy S20 fe), I use games on the sd card, I have the latest version of Daijisho and version 0.9.5 of melonDS (which is here on github ) I have no idea what could be going on because here with me everything works normally...

**Reply by jackalopy** [reply]
It might have something to have with the way the path to the rom is written. Maybe you could show us how your paths are formatted in daijisho?

**Reply by EgoBizarro** [reply]
@jackalopy: Sorry, I'm not sure I understand what you mean. 
Daijisho 'wrote' the path for me, I just picked the folder. 
The path is /tree/primary:ROMs/Nintendo DS
I have other roms in the same ROMs folder and they work fine. e.g.:
/tree/primary:ROMs/Nintendo Gamecube

**Reply by jackalopy** [reply]
Today I dived into the MelonDS source code (not an android dev). Here is the solution I've found:

```
-a me.magnum.melonds.LAUNCH_ROM
-n me.magnum.melonds/.ui.emulator.EmulatorActivity
-e uri "{file.uri}"
```

I hope this helps you guys

**Reply by EgoBizarro** [reply]
That's it! With that it works great for me, thanks a lot! :D
To my surprise, I can even boot DSi games via Daijisho - obviously, I have to have DSi selected as default system in melonDS for them to work. BTW, does anyone here know if there's any downside to having DSi set as default system? (since it still says '(experimental)'.

**Reply by SoumyBhow** [reply]
Interesting. The fix works on Daijsho but not on Pegasus frontend. After setting everything to one line and replacing file.uri with file.documenturi, MelonDS says it coudn't find the ROM. Has anyone else managed to get this to work on Pegasus?

**Reply by jackalopy** [reply]
If what I'm seeing here https://pegasus-frontend.org/docs/user-guide/meta-files/ is right, maybe you should use {file.uri} anyway.
Why would you change it to {file.documenturi}?

**Reply by SoumyBhow** [reply]
> If what I'm seeing here https://pegasus-frontend.org/docs/user-guide/meta-files/ is right, maybe you should use {file.uri} anyway. Why would you change it to {file.documenturi}?

It's cause before, file.uri was an attempt by the dev to fix launching from emulators that use SAF (like PPSSPP, M64PFZ, etc). However that didn't work well, so the dev implemented the new way as file.documenturi. Also I've already tried file.uri but with no success.

**Reply by gtaforever00** [reply]
So looking into Pegasus and the issues with MelonDS launching, it seems it's related to how Pegasus encodes the uri. I'm not sure if Pegasus or MelonDS needs changed. It seems Pegasus encodes Unreserved Characters that RFC2396 states the following. 

```
Unreserved characters can be escaped without changing the semantics
   of the URI, but this should not be done unless the URI is being used
   in a context that does not allow the unescaped character to appear.
```

http://www.faqs.org/rfcs/rfc2396.html

Other emulators seem fine with this so I'm not sure who would need to fix the issue. 

TLDR;
Remove any unreserved characters from your file name in Pegasus and rom file and it works.

unreserved characters

```
"-" | "_" | "." | "!" | "~" | "*" | "'" | "(" | ")"
```

**Comment by ciaskohansen** [comment]
Hello, any progress with its use in frontends? I'm on Android 13 and I use Arc Browser and following the last comments there is no way to launch it.

**Comment by lauhoitsun2018-a11y** [comment]
collection: NDS
extensions: nds, zip, 7z, dsi
launch: am start --user 0
 -a android.intent.action.VIEW
 -n me.magnum.melondualds/me.magnum.melonds.ui.emulator.EmulatorActivity
 -d "{file.uri}"

it works for me
